Add weather widget

zMoooooritz opened the pull request:
This is my take on a new widget, the weather widget.

Overall I'm more or less happy with the way it currently works but it needs some improvements:

1. When opening a deck that contains this widget the new deck is rendered until this widget is next, it takes then 0.1 - 3 seconds to fetch the data from the internet and render the remaining widgets. In case the data is cached or it is not the first update its faster otherwise it takes up to 3 seconds. Any ideas how to solve this problem (most importantly the intial update)?
2. I have done the serialized part of the images before you have implemented proper relative paths from the location of `deckmaster`. Would we want to keep this or add all the required images to the assets and load them as needed.
3. Saving and loading the images would probably allow for better adjustment to the button size.

muesli commented on widget_weather.go:
```diff
@@ -0,0 +1,136 @@
+	case "f", "fahrenheit":
+		return "&u"
+	case "c", "celcius":
+		return "&u"
```
This seems to return `&u` for both Fahrenheit & Celsius?

zMoooooritz replied:
Fixed that one dont know how that happened ...

muesli commented on widget_weather.go:
```diff
@@ -0,0 +1,136 @@
+
+	wttr := strings.Split(string(body), " ")
+	cond := wttr[0]
+	temp := strings.Replace(wttr[1], "+", "", 1)
```
We should check that `wttr` actually contains two items before accessing `wttr[1]`.

muesli replied:
Just did that.

muesli commented:
Just pushed a small change that fetches the data in the background, and cleans up the weather loading/parsing a bit. Let me know what you think!

muesli commented:
Last but not least: we need to update the README, and maybe even put an example into `main.deck`.

zMoooooritz commented:
> Another day, another widget  Thank you for all your contributions!

:D but I think I'm know starting to run out of ideas

> I'd say add the assets to the repo. This is such an essential widget and (unlike the Emoji keyboard) doesn't spam the repo with all kinds of images.

I have changed your assets and added some more icons.
In case you dont like them just drop my last commit.

> Just pushed a small change that fetches the data in the background, and cleans up the weather loading/parsing a bit. Let me know what you think!

Would have loved to try that myself to learn some more non-basic go := but thank you that looks great!

> Last but not least: we need to update the README, and maybe even put an example into `main.deck`.

Yes I have done that.

Furthermore I added a new attribute `bool flatten` in case it is set all pixels with an alpha above a threshold will be set to the widget `color` all the others will be transparent.
Made this change since I like to have my foreground in one color ;)

muesli commented:
> :D but I think I'm know starting to run out of ideas

Should you be looking for one, let me know. I'd love to see an audio mixer deck, integration with OBS, VSCode, etc.

> I have changed your assets and added some more icons.
> In case you dont like them just drop my last commit. (I just saw that I change the type of `alpha_threshold` in this commit in case you drop the commit you need to keep this change)

I need to check them out properly, but some seem a bit problematic at first glance: e.g. the Cloud/Snow icon has black snowflakes? :laughing:

> Would have loved to try that myself to learn some more non-basic go := but thank you that looks great!

Darn, I'm really sorry. I hope there's still something you can take away from looking at how I approached it.

> > Last but not least: we need to update the README, and maybe even put an example into `main.deck`.
> 
> Yes I have done that.

Awesome!

> Furthermore I added a new attribute `bool flatten` in case it is set all pixels with an alpha above a threshold will be set to the widget `color` all the others will be transparent.
> Made this change since I like to have my foreground in one color ;)

Maybe that's an option we actually want to add to the `ButtonWidget` instead?

zMoooooritz commented:
> Should you be looking for one, let me know. I'd love to see an audio mixer deck, integration with OBS, VSCode, etc.

Sure, the audio mixer and OBS integration sound good.
Maybe we could create a list of ideas / todos to provide a good starting point for new contributors.

> I need to check them out properly, but some seem a bit problematic at first glance: e.g. the Cloud/Snow icon has black snowflakes? 

In case we can't agree / have different thought on the icons, we could maybe let the user decide on the icons.
So there is a attribute like `iconDir` and each user can place their icons inside there and just needs to name them correctly?

> Darn, I'm really sorry. I hope there's still something you can take away from looking at how I approached it.

No worries. That is also a good way to learn best practices of a new language! :+1: 

> Maybe that's an option we actually want to add to the `ButtonWidget` instead?

Sounds good. I will have a look at it in the evening.

muesli commented:
> Sure, the audio mixer and OBS integration sound good.
> Maybe we could create a list of ideas / todos to provide a good starting point for new contributors.

Agreed. Let's just create a bunch of issues with ideas here.

> In case we can't agree / have different thought on the icons, we could maybe let the user decide on the icons.
> So there is a attribute like `iconDir` and each user can place their icons inside there and just needs to name them correctly?

Something like an icon theme is certainly an option.

> Edit: wttr.in is down right now ... it does return `Unknown location;` for our query which results in deckmaster quitting.
> Maybe we should overthink the line `return "", fmt.Errorf("unknown location: %s", w.location)` maybe we should just print the error to stderr and draw an empty widget?

I agree, quitting is probably too harsh.

zMoooooritz commented:
> Something like an icon theme is certainly an option.

I hope I understood that corretly and set it up the way you wanted :thinking:
As of now openmoji is the default theme since no other complete theme exists

> I agree, quitting is probably too harsh.

Updated the relevant part to now just show an empty widget

muesli commented on widget_weather.go:
```diff
@@ -0,0 +1,229 @@
+		return w.render(dev, nil)
+	}
+
+	weatherIcon := weatherImage("assets/weather/" + w.theme + "/" + iconName + ".png")
```
Instead of manually concatenating a path here, we should use `filepath.Join("assets", "weather", w.theme, iconName+".png")`.

muesli commented:
> I hope I understood that corretly and set it up the way you wanted thinking
> As of now openmoji is the default theme since no other complete theme exists

This probably means we should remove the now obsolete images in `assets/weather/*.png`?

I think the theme handling needs to be a bit cleverer. It's good that we have a default/fallback we can embed in the binary, but I would imagine the user-configured themes need to be loaded from disk (e.g. from `~/.local/share/deckmaster/themes`), so one doesn't have to rebuild the binary just to switch themes.

> Updated the relevant part to now just show an empty widget

Fantastic!

zMoooooritz commented:
> This probably means we should remove the now obsolete images in `assets/weather/*.png`?
> 
> I think the theme handling needs to be a bit cleverer. It's good that we have a default/fallback we can embed in the binary, but I would imagine the user-configured themes need to be loaded from disk (e.g. from `~/.local/share/deckmaster/themes`), so one doesn't have to rebuild the binary just to switch themes.

I have moved the new widgets to `assets/weather` and deleted the ones that have been there before.
I hope this fulfills your requirements to load the images from `~/.local/share/deckmaster/themes` and I have not forgotten to catch any errors

There are two things I am not sure about as of yet:

1. Do we want to support other file types e.g. `jpg` and maybe even more? I would have no clue how to do this properly :confused: 
2. At least for me the `widget_weather` gets updated every interval (so 10 times per second). I cant pinpoint the reason for that as of now or did you intend to have it that way?

Feel free to change things. I call it a day :smile:

muesli commented on widget.go:
```diff
@@ -160,6 +164,21 @@ func loadImage(path string) (image.Image, error) {
+		return nil, err
+	}
+	var image image.Image
+	image, err = loadImage(path)
```
Since the signatures of the functions match here, we can simply:

```go
return loadImage(path)
```

muesli commented on widget_button.go:
```diff
@@ -57,6 +66,9 @@ func (w *ButtonWidget) Update(dev *streamdeck.Device) error {
 	height := size - (margin * 2)
 	img := image.NewRGBA(image.Rect(0, 0, size, size))
 
+	if w.icon != nil && w.flatten {
```
I think this should be done in `NewButtonWidget`, so it only gets flattened once, not on every call of `Update`.

zMoooooritz replied:
In case we move the call of `flattenImage` into `NewButtonWidget` this wont apply for the weather widget.
Since the weather image does only construct a `ButtonWidget` and not call `NewButtonWidget`

zMoooooritz commented:
> 2. At least for me the `widget_weather` gets updated every interval (so 10 times per second). I cant pinpoint the reason for that as of now or did you intend to have it that way?

> In case we move the call of `flattenImage` into `NewButtonWidget` this wont apply for the weather widget.
> Since the weather widget does only construct a `ButtonWidget` and not call `NewButtonWidget`

These two things are closly connected.
With the current implementation of the `widget_weather` each `Update` a new `ButtonWidget` is constructed. Therefore the `lastUpdate` will never be set since the next iteration creates a new `ButtonWiget` with a new `lastUpdate`. Hence the `WeatherWidget` is updated in each iteration.

One possible solution I can think of rn is to take the `Update` of the `ButtonWidget` and place it somewhere in `widget.go` and call it from `ButtonWIdget` and `WeatherWidget`.

Since the number of files in the root-dir is becoming rather long we should maybe create a dir `widgets` and move all the widgets in there?

muesli commented:
@zMoooooritz Ok, just pushed some fixes that address outstanding issues:

1. Embed `ButtonWidget` in `WeatherWidget` and `RecentWindowWidget`. This means we also inherit their `BaseWidget`s and the `UpdateRequired` check works correctly again.
2. Inherit `ButtonWidget`'s settings in `WeatherWidget` and `RecentWindowWidget` without having to redeclare them.

zMoooooritz commented on widget_weather.go:
```diff
@@ -0,0 +1,226 @@
+
+// NewWeatherWidget returns a new WeatherWidget.
+func NewWeatherWidget(bw *BaseWidget, opts WidgetConfig) (*WeatherWidget, error) {
+	bw.setInterval(opts.Interval, 60000)
```
As far as I tested this does not change the update interval... I'm still debugging but there seems to be something wrong

muesli replied:
I see what's happening: `NewButtonWidget` overrides this by calling `bw.SetInterval` itself. Fixing.

muesli replied:
Ok, pushed a fix. This also fixes not drawing the background initially while waiting for the first weather data response. I'm not too fond of how I fixed it, however. Will think of a nicer solution.

muesli commented:
Ok, fairly happy with this now. It's gotten a lot bigger than anticipated, and before we keep piling on more changes on top here, I figured it's time to merge.

Thank you, once again!

zMoooooritz commented:
> Ok, fairly happy with this now. It's gotten a lot bigger than anticipated, and before we keep piling on more changes on top here, I figured it's time to merge.

:+1: 

> Thank you, once again!

Nothing to thank for! Thank YOU for creating this awesome project, allowing me contribute and helping me with my first steps in go! :100:
